File: leetcode/144_Binary_Tree_Preorder_Traversal.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void preorder(TreeNode* root, vector<int> &res){
        if(root==nullptr)return;
        res.push_back(root->val);
        preorder(root->left,res);
        preorder(root->right,res);
        
    }
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int>res;
        preorder(root, res);
        return res;
    }
};
int main(){
    Solution s;
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    vector<int> result = s.preorderTraversal(root);
    cout<<"Preorder Traversal: ";
    for(int val : result){
        cout<<val<<" ";
    }
    cout<<endl;
    return 0;
}

File: leetcode/144_Binary_Tree_Preorder_Traversal.java
import java.util.ArrayList;
public class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {
         this.val = val;
         this.left = left;
         this.right = right;
     }
 }
 
class Solution {
    ArrayList<Integer> preorder= new ArrayList<>();
    public void BST( TreeNode root){
        if(root==null)return;
        preorder.add(root.val);
        BST(root.left);
        BST(root.right);
    }
    public List<Integer> preorderTraversal(TreeNode root) {
        BST(root);
        return preorder;
    }
}
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) return res;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            res.add(node.val);
            if (node.right != null) stack.push(node.right);
            if (node.left != null) stack.push(node.left);
        }
        return res;
    }
}